Lib/test/test_fileio.py

```diff
@@ -359,6 +362,85 @@ def testErrnoOnClosedReadinto(self, f):
         a = array('b', b'x'*10)
         f.readinto(a)
 
+    @strace_helper.requires_strace()
+    def test_syscalls_read(self):
+        """Check that the set of system calls produced by the I/O stack is what
+        is expected for various read cases.
+
+        It's expected as bits of the I/O implementation change, this will need
+        to change. The goal is to catch changes that unintentionally add
+        additional systemcalls (ex. additional fstat calls has been an issue).
+        """
+        self.f.write(b"Hello, World!")
+        self.f.close()
+
+        # "open, read, close" file with default options.
+        calls = strace_helper.get_syscalls(
+            f"""
+            f = open('{TESTFN}')
+            f.read()
+            f.close()
+            """,
+            _strace_flags
+        )
+
+        # There are a number of related syscalls used to implement `fstat` in
+        # a libc (ex. fstat, newfstatat). Default to fstat, but if we see the
+        # system using a variant, allow that
+        fstat = next((sc for sc in calls if 'fstat' in sc), 'fstat')
+
+        readall_calls = ['openat', fstat, 'ioctl', 'lseek', 'read',
```

Comment:
I agree this is a a potentially fragile point, and lots could lead to a test failure even though the cpython implementation is as desired (ex. cpython `_io`, cpython `_pyio`, glibc, and musl differ around Buffered I/O). There are a lot of system calls which _could_ happen as part of the Python standard file opening process which I want to validate don't ideally. Just saying "one stat", "two reads" "no ioctl" doesn't quite feel like it could help catch these changes.

Motivating examples I would like to be able to write behavior tests around:
1. CPython's `open()` builtin currently calls `isatty` to check if the file it just opened is "interactive" to determine buffered mode. If buffered, a `BufferedIO` is constructed which additionally wants to have an accurate position inside the file, and calls `lseek`, sometimes at `__init__` (inside open), but sometimes when other I/O operations happen (`read`, `write`). The buffering behavior is visible, and documented, around stdout (https://docs.python.org/3/library/sys.html#sys.stdout) with a special environment variable to change it `PYTHONUNBUFFERED`. `stderr` has a different behavior than stdout. For this PR, I needed to do a `flush=True` in `print()` because the test `python` has its output redirected to a file which changes the behavior of the standard library making there be one `write()` system call at interpreter shutdown instead of calls when the `print()` statements are run. The `isatty` python check results in an `ioctl` system call (full stack at [0]).
1. https://github.com/python/cpython/pull/121315 in the initial implementation added a `fstat` to every `read` call to test "is this a pipe" (https://github.com/python/cpython/pull/121315/commits/108e65bc3a2a142ed2bb347e5222a626c477d93c). That would fairly substantially reduce read performance. If I just test for `at most one fstat` over lifetime, could catch. But what if it was an attribute which `ioctl` is used to get? The performance slowdown was happening because of lots large allocation + de-allocation (gh-121313), which isn't a FD system call.

In some ways this feels like it interacts a bit with "what is a supported platform" (https://peps.python.org/pep-0011/). I can defer for now and count specific system calls (1 fstat over life of file, no ioctls, etc.) but worry it leaves some substantial gaps. I'm really not sure what the right set of tradeoffs is for long term maintenance + platform porting and needing to update tests like this one.


[0] 
1. `ioctl(TCGETS)` (linux)
1. `tcgettattr` (glibc)
1. `isatty` (glibc)
1. `io.isatty` (cpython)
1. `FileIO.isatty` (cpython)
1. `open()` (cpython)